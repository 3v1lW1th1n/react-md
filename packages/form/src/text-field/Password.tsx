import React, {
  CSSProperties,
  FC,
  forwardRef,
  ReactNode,
  useCallback,
  useState,
} from "react";
import cn from "classnames";
import { Button } from "@react-md/button";
import { FontIcon } from "@react-md/icon";
import { bem } from "@react-md/theme";
import { Omit, WithForwardedRef } from "@react-md/utils";

import TextField, { TextFieldProps } from "./TextField";

export interface PasswordProps
  extends Omit<TextFieldProps, "type" | "rightChildren"> {
  /**
   * The icon to use to toggle the visibility of the password by changing the
   * input type to text temporarily.
   */
  visibilityIcon?: ReactNode;

  /**
   * An optional style to apply to the visibility toggle button.
   */
  visibilityStyle?: CSSProperties;

  /**
   * An optional classname to apply to the visibility toggle button.
   */
  visibilityClassName?: string;

  /**
   * An optional `aria-label` to apply to the visibility toggle button.
   */
  visibilityLabel?: string;

  /**
   * Boolean if the visibility toggle feature should be disabled.
   */
  disableVisibility?: boolean;
}

type WithRef = WithForwardedRef<HTMLInputElement>;
type DefaultProps = Required<
  Pick<
    PasswordProps,
    "visibilityIcon" | "visibilityLabel" | "disableVisibility"
  >
>;
type WithDefaultProps = PasswordProps & DefaultProps & WithRef;

const block = bem("rmd-password");

/**
 * This component is a simple wrapper of the `TextField` that can only be rendered
 * for password inputs. There is built-in functionality to be able to temporarily
 * show the password's value by swapping the `type` to `"text"`.
 */
const Password: FC<PasswordProps & WithRef> = providedProps => {
  const {
    className,
    inputClassName,
    forwardedRef,
    visibilityIcon,
    disableVisibility,
    visibilityStyle,
    visibilityClassName,
    visibilityLabel,
    ...props
  } = providedProps as WithDefaultProps;
  const [type, setType] = useState<"password" | "text">("password");
  const toggle = useCallback(() => {
    setType(prevType => (prevType === "password" ? "text" : "password"));
  }, []);

  return (
    <TextField
      {...props}
      className={cn(block({ offset: !disableVisibility }), className)}
      inputClassName={cn(
        block("input", { offset: !disableVisibility }),
        inputClassName
      )}
      ref={forwardedRef}
      type={type}
      isRightAddon={false}
      rightChildren={
        !disableVisibility && (
          <Button
            id={`${props.id}-password-toggle`}
            aria-label={visibilityLabel}
            buttonType="icon"
            onClick={toggle}
            style={visibilityStyle}
            className={cn(block("toggle"), visibilityClassName)}
          >
            {visibilityIcon}
          </Button>
        )
      }
    />
  );
};

const defaultProps: DefaultProps = {
  visibilityIcon: <FontIcon>remove_red_eye</FontIcon>,
  visibilityLabel: "Temporarily show password",
  disableVisibility: false,
};

Password.defaultProps = defaultProps;

if (process.env.NODE_ENV !== "production") {
  Password.displayName = "Password";

  let PropTypes = null;
  try {
    PropTypes = require("prop-types");
  } catch (e) {}

  if (PropTypes) {
    Password.propTypes = {
      id: PropTypes.string.isRequired,
      visibilityIcon: PropTypes.node,
      visibilityStyle: PropTypes.object,
      visibilityClassName: PropTypes.string,
      visibilityLabel: PropTypes.string,
      disableVisibility: PropTypes.bool,
    };
  }
}

export default forwardRef<HTMLInputElement, PasswordProps>((props, ref) => (
  <Password {...props} forwardedRef={ref} />
));
