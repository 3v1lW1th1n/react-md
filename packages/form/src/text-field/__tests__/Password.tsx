import React from "react";
import { render, fireEvent } from "@testing-library/react";

import Password from "../Password";

const getById = (id: string) => {
  const el = document.getElementById(id);
  if (!el) {
    throw new Error();
  }

  return el;
};

describe("Password", () => {
  it("should render correctly", () => {
    const { container, rerender } = render(<Password id="password" />);

    expect(container).toMatchSnapshot();

    rerender(
      <Password id="password" label="Password" placeholder="Password..." />
    );
    expect(container).toMatchSnapshot();
  });

  it("should switch to a text input when the visibility button is clicked", () => {
    render(<Password id="password" label="Label" />);

    const password = getById("password");
    const toggle = getById("password-password-toggle");

    expect(password).toMatchSnapshot();
    expect(password.getAttribute("type")).toBe("password");

    fireEvent.click(toggle);
    expect(password).toMatchSnapshot();
    expect(password.getAttribute("type")).toBe("text");

    fireEvent.click(toggle);
    expect(password).toMatchSnapshot();
    expect(password.getAttribute("type")).toBe("password");
  });

  it("should not render the password toggle button if the disableVisibility prop is enabled", () => {
    const { container, rerender } = render(<Password id="password" />);
    const getToggle = () => getById("password-password-toggle");

    expect(container).toMatchSnapshot();
    expect(getToggle).not.toThrow();

    rerender(<Password id="password" disableVisibility />);
    expect(container).toMatchSnapshot();
    expect(getToggle).toThrow();
  });
});
