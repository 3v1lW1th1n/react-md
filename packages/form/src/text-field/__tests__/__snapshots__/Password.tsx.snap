// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Password should not render the password toggle button if the disableVisibility prop is enabled 1`] = `
<div>
  <div
    class="rmd-text-field-container rmd-text-field-container--outline rmd-text-field-container--hoverable rmd-text-field-container--outline-right rmd-password rmd-password--offset"
  >
    <input
      class="rmd-text-field rmd-password__input rmd-password__input--offset"
      id="password"
      type="password"
    />
    <button
      aria-label="Temporarily show password"
      class="rmd-button rmd-button--icon rmd-password__toggle"
      id="password-password-toggle"
      type="button"
    >
      <i
        aria-hidden="true"
        class="rmd-icon rmd-icon--font material-icons"
      >
        remove_red_eye
      </i>
      <span
        class="rmd-ripple-container"
      />
    </button>
  </div>
</div>
`;

exports[`Password should not render the password toggle button if the disableVisibility prop is enabled 2`] = `
<div>
  <div
    class="rmd-text-field-container rmd-text-field-container--outline rmd-text-field-container--hoverable rmd-password"
  >
    <input
      class="rmd-text-field rmd-password__input"
      id="password"
      type="password"
    />
  </div>
</div>
`;

exports[`Password should render correctly 1`] = `
<div>
  <div
    class="rmd-text-field-container rmd-text-field-container--outline rmd-text-field-container--hoverable rmd-text-field-container--outline-right rmd-password rmd-password--offset"
  >
    <input
      class="rmd-text-field rmd-password__input rmd-password__input--offset"
      id="password"
      type="password"
    />
    <button
      aria-label="Temporarily show password"
      class="rmd-button rmd-button--icon rmd-password__toggle"
      id="password-password-toggle"
      type="button"
    >
      <i
        aria-hidden="true"
        class="rmd-icon rmd-icon--font material-icons"
      >
        remove_red_eye
      </i>
      <span
        class="rmd-ripple-container"
      />
    </button>
  </div>
</div>
`;

exports[`Password should render correctly 2`] = `
<div>
  <div
    class="rmd-text-field-container rmd-text-field-container--outline rmd-text-field-container--hoverable rmd-text-field-container--label rmd-text-field-container--outline-right rmd-password rmd-password--offset"
  >
    <label
      class="rmd-label rmd-floating-label"
      for="password"
    >
      Password
    </label>
    <input
      class="rmd-text-field rmd-text-field--floating rmd-password__input rmd-password__input--offset"
      id="password"
      placeholder="Password..."
      type="password"
    />
    <button
      aria-label="Temporarily show password"
      class="rmd-button rmd-button--icon rmd-password__toggle"
      id="password-password-toggle"
      type="button"
    >
      <i
        aria-hidden="true"
        class="rmd-icon rmd-icon--font material-icons"
      >
        remove_red_eye
      </i>
      <span
        class="rmd-ripple-container"
      />
    </button>
  </div>
</div>
`;

exports[`Password should switch to a text input when the visibility button is clicked 1`] = `
<input
  class="rmd-text-field rmd-text-field--floating rmd-password__input rmd-password__input--offset"
  id="password"
  type="password"
/>
`;

exports[`Password should switch to a text input when the visibility button is clicked 2`] = `
<input
  class="rmd-text-field rmd-text-field--floating rmd-password__input rmd-password__input--offset"
  id="password"
  type="text"
/>
`;

exports[`Password should switch to a text input when the visibility button is clicked 3`] = `
<input
  class="rmd-text-field rmd-text-field--floating rmd-password__input rmd-password__input--offset"
  id="password"
  type="password"
/>
`;
